### PR TITLE
Insider Integration

### PR DESCRIPTION
**Pull request by emreisik95**
Insider Authentication, Update User Profile, and Track Event feature added.

## Testing

_Include any additional information about the testing you have completed to
ensure your changes behave as expected. For a speedy review, please check
any of the tasks you completed below during your testing._

- [x] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [x] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
- [ ] [Segmenters] Tested in the staging environment

### CODE REVIEW

**Comment by joe-ayoub-segment** [comment]
Hi @emreisik95 , thanks for raising this PR. 
Looks like it's for a new Integration :). Would you be able to schedule a call with me so we can run through the use-cases that customers will use this Integration for please? I'd also like to get a picture of the types of data which will be forwarded from Segment to this new Integration, as well as the identifiers which will be used. 

You can schedule a call with me using [this Calendly link](http://calendly.com/joe_ayoub/integration-workshop). 

Kind regards,
Joe

File: packages/destination-actions/src/destinations/insider/updateUserProfile/index.ts [hunk 1]
```diff
@@ -0,0 +1,143 @@
+import type { ActionDefinition } from '@segment/actions-core'
+import type { Settings } from '../generated-types'
+import type { Payload } from './generated-types'
+import { userProfilePayload, API_BASE, UPSERT_ENDPOINT } from '../insider-helpers'
+
+const action: ActionDefinition<Settings, Payload> = {
+  title: 'Update User Profile',
+  description: '',
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/insider/updateUserProfile/index.ts:8]: add descrtiption

File: packages/destination-actions/src/destinations/insider/updateUserProfile/index.ts [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+      allowNull: true,
+      description: 'Age of a user.',
+      default: {
+        '@path': '$.traits.age'
+      }
+    },
+    birthday: {
+      label: 'Birthday',
+      type: 'string',
+      allowNull: true,
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/insider/updateUserProfile/index.ts:23]: remove

File: packages/destination-actions/src/destinations/insider/updateUserProfile/index.ts [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+const action: ActionDefinition<Settings, Payload> = {
+  title: 'Update User Profile',
+  description: '',
+  defaultSubscription: 'type = "identify"',
+  fields: {
+    age: {
+      label: 'Age',
+      type: 'number',
+      allowNull: true,
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/insider/updateUserProfile/index.ts:14]: remove

File: packages/destination-actions/src/destinations/insider/updateUserProfile/index.ts [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+      description: 'User’s birthday',
+      default: {
+        '@path': '$.traits.birthday'
+      }
+    },
+    email: {
+      label: 'Email',
+      type: 'string',
+      allowNull: true,
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/insider/updateUserProfile/index.ts:32]: remove

File: packages/destination-actions/src/destinations/insider/updateUserProfile/index.ts [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+        '@path': '$.traits.email'
+      }
+    },
+    firstName: {
+      label: 'First Name',
+      description: 'First name of a user.',
+      type: 'string',
+      allowNull: true,
+      default: {
+        '@path': '$.traits.firstName'
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/insider/updateUserProfile/index.ts:44]:
```suggestion
        '@path': '$.traits.first_name'
```

File: packages/destination-actions/src/destinations/insider/updateUserProfile/index.ts [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+      description: 'Email address of a user.',
+      default: {
+        '@path': '$.traits.email'
+      }
+    },
+    firstName: {
+      label: 'First Name',
+      description: 'First name of a user.',
+      type: 'string',
+      allowNull: true,
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/insider/updateUserProfile/index.ts:42]: remove

File: packages/destination-actions/src/destinations/insider/updateUserProfile/index.ts [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+      default: {
+        '@path': '$.traits.firstName'
+      }
+    },
+    gender: {
+      label: 'Gender',
+      description: 'Gender of a user.',
+      type: 'string',
+      allowNull: true,
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/insider/updateUserProfile/index.ts:51]: remove

File: packages/destination-actions/src/destinations/insider/updateUserProfile/index.ts [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+        '@path': '$.traits.gender'
+      }
+    },
+    lastName: {
+      label: 'Last Name',
+      description: 'Last name of a user.',
+      type: 'string',
+      allowNull: true,
+      default: {
+        '@path': '$.traits.lastName'
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/insider/updateUserProfile/index.ts:62]:
```suggestion
        '@path': '$.traits.last_name'
```

File: packages/destination-actions/src/destinations/insider/updateUserProfile/index.ts [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+      allowNull: true,
+      default: {
+        '@path': '$.traits.gender'
+      }
+    },
+    lastName: {
+      label: 'Last Name',
+      description: 'Last name of a user.',
+      type: 'string',
+      allowNull: true,
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/insider/updateUserProfile/index.ts:60]: remove

File: packages/destination-actions/src/destinations/insider/updateUserProfile/index.ts [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+      default: {
+        '@path': '$.traits.lastName'
+      }
+    },
+    phone: {
+      label: 'Phone Number',
+      description: 'Phone number of a user.',
+      type: 'string',
+      allowNull: true,
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/insider/updateUserProfile/index.ts:69]: remove

File: packages/destination-actions/src/destinations/insider/updateUserProfile/index.ts [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+      default: {
+        '@path': '$.traits.phone'
+      }
+    },
+    uuid: {
+      label: 'UUID',
+      type: 'string',
+      allowNull: false,
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/insider/updateUserProfile/index.ts:77]: remove

File: packages/destination-actions/src/destinations/insider/updateUserProfile/index.ts [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+      description: 'User unique id.',
+      default: {
+        '@path': '$.userId'
+      }
+    },
+    segment_anonymous_id: {
+      label: 'Anonymous Id',
+      type: 'string',
+      allowNull: false,
+      description: 'Anonymous user id.',
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/insider/updateUserProfile/index.ts:87]: update description to indicate that userId or anonymousId is required.

File: packages/destination-actions/src/destinations/insider/updateUserProfile/index.ts [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+      allowNull: true,
+      description: 'City',
+      default: {
+        '@path': '$.traits.address.city'
+      }
+    },
+    country: {
+      label: 'Country',
+      type: 'string',
+      allowNull: true,
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/insider/updateUserProfile/index.ts:104]: remove

File: packages/destination-actions/src/destinations/insider/updateUserProfile/index.ts [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+      description: 'Country',
+      default: {
+        '@path': '$.traits.address.country'
+      }
+    },
+    emailOptin: {
+      label: 'Email Optin',
+      type: 'string',
+      allowNull: true,
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/insider/updateUserProfile/index.ts:113]: remove

File: packages/destination-actions/src/destinations/insider/updateUserProfile/index.ts [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+    smsOptin: {
+      label: 'SMS Optin',
+      type: 'string',
+      allowNull: true,
+      description: 'SMS optin.'
+    },
+    whatsappOptin: {
+      label: 'Whatsapp Optin',
+      type: 'string',
+      allowNull: true,
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/insider/updateUserProfile/index.ts:125]: remove

File: packages/destination-actions/src/destinations/insider/updateUserProfile/index.ts [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+      description: 'Whatsapp optin.'
+    },
+    language: {
+      label: 'Language',
+      description: "The user's preferred language.",
+      type: 'string',
+      allowNull: true
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/insider/updateUserProfile/index.ts:132]: remove

**Review by joe-ayoub-segment — CHANGES_REQUESTED** [review]

**Review by joe-ayoub-segment — CHANGES_REQUESTED** [review]
Hi team, 
Thanks for taking the time to catch up just now. 

Main items were: 
1. Add presets
2. Provide better descriptions
3. correct some default field mappings

During our call we didn't look at unit tests. I'll run them now to make sure they pass, and will review them to check how much code coverage there is.

**Comment by emreisik95** [comment]
Thank you for your help and code review @joe-ayoub-segment. 

@recep-insider @esezerguven will check the comments and will make the requested changes.

**Comment by emreisik95** [comment]
> Hi team, Thanks for taking the time to catch up just now.
> 
> Main items were:
> 
> 1. Add presets
> 2. Provide better descriptions
> 3. correct some default field mappings
> 4. Add additional fields to let customer select if email and phone should be treated as identifiers
> 
> During our call we didn't look at unit tests. I'll run them now to make sure they pass, and will review them to check how much code coverage there is.

Can you send us an example of the 4th item?

**Comment by joe-ayoub-segment** [comment]
Hi @emreisik95 , this is what I was thinking (see attached video). Totally up to you if this makes sense for your platform or not. You are the subject matter experts!


https://user-images.githubusercontent.com/45374896/224013794-4b73bb1e-417b-4d78-b0c1-e107effdc632.mov

**Comment by recep-insider** [comment]
@joe-ayoub-segment hello, when I try to run unit tests on my local I got these errors. 
image

Do you any suggestion? Thanks.

**Comment by joe-ayoub-segment** [comment]
Hi @recep-insider have you run `yarn && yarn build` first?

**Comment by recep-insider** [comment]
> Hi @recep-insider have you run `yarn && yarn build` first?

Yes but when I run this command then I got another errors.
image

**Comment by joe-ayoub-segment** [comment]
Are you able to try this first please: 
`yarn --ignore-optional && yarn build'

**Comment by recep-insider** [comment]
> 

still same

**Comment by joe-ayoub-segment** [comment]
Would you like to join a call ? Maybe I can help you debug.
https://twilio.zoom.us/j/94525165341?pwd=WTJyaWd0WVByOG11NURqWS9UdXVWUT09

**Comment by recep-insider** [comment]
@joe-ayoub-segment Hello, I think we are ready to CR again.

**Comment by joe-ayoub-segment** [comment]
Hi @recep-insider I'll schedule this for a review. Thank you!

File: packages/destination-actions/src/destinations/insider/updateUserProfile/index.ts [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+        "User's unique user ID. UUID should be string and it is used as identifier when sending data to Insider. Either Anonymous ID or UUID is mandatory to send data.",
+      default: {
+        '@path': '$.userId'
+      }
+    },
+    segment_anonymous_id: {
+      label: 'Anonymous Id',
+      type: 'string',
+      description:
+        'Segment Anonymous ID. It is used as identifier when sending data to Insider. Either Anonymous ID or UUID is mandatory to send data.',
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/insider/updateUserProfile/index.ts:92]:
```suggestion
        'An Anonymous Identifier. The Anonymous Id string is used as identifier when sending data to Insider. Anonymous Id is required if the UUID field is empty.',
```

File: packages/destination-actions/src/destinations/insider/trackEvent/index.ts [hunk 18]
```diff
@@ -0,0 +1,374 @@
...
+  title: 'Track Event',
+  description: 'Records events in Insider',
+  defaultSubscription:
+    'type = "track" and event != "Order Completed" and event != "Cart Viewed" and event != "Checkout Viewed"',
+  fields: {
+    uuid: {
+      label: 'User ID',
+      type: 'string',
+      description:
+        "User's unique user ID. UserID should be string and it is used as identifier when sending data to Insider. Either Anonymous ID or UUID is mandatory to send data",
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/insider/trackEvent/index.ts:16]:
```suggestion
        "User's unique identifier. The UUID string is used as identifier when sending data to Insider. UUID is required if the Anonymous Id field is empty.",
```

File: packages/destination-actions/src/destinations/insider/trackEvent/index.ts [hunk 19]
```diff
@@ -0,0 +1,374 @@
...
+import { API_BASE, UPSERT_ENDPOINT, sendTrackEvent } from '../insider-helpers'
+
+const action: ActionDefinition<Settings, Payload> = {
+  title: 'Track Event',
+  description: 'Records events in Insider',
+  defaultSubscription:
+    'type = "track" and event != "Order Completed" and event != "Cart Viewed" and event != "Checkout Viewed"',
+  fields: {
+    uuid: {
+      label: 'User ID',
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/insider/trackEvent/index.ts:13]:
```suggestion
      label: 'UUID',
```

File: packages/destination-actions/src/destinations/insider/index.ts [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+/** used in the quick setup */
+const presets: DestinationDefinition['presets'] = [
+  {
+    name: 'Identify Calls',
+    subscribe: 'type = "identify"',
+    partnerAction: 'updateUserProfile',
+    mapping: defaultValues(updateUserProfile.fields)
+  },
+  {
+    name: 'Track Calls',
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/insider/index.ts:17]:
```suggestion
    name: 'Record user event to Insider',
```

File: packages/destination-actions/src/destinations/insider/index.ts [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+import type { Settings } from './generated-types'
+
+import updateUserProfile from './updateUserProfile'
+import trackEvent from './trackEvent'
+import { defaultValues } from '@segment/actions-core'
+
+/** used in the quick setup */
+const presets: DestinationDefinition['presets'] = [
+  {
+    name: 'Identify Calls',
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/insider/index.ts:11]:
```suggestion
    name: 'Create or update a user profile',
```

File: packages/destination-actions/src/destinations/insider/updateUserProfile/index.ts [hunk 22]
```diff
@@ -0,0 +1,142 @@
+import type { ActionDefinition } from '@segment/actions-core'
+import type { Settings } from '../generated-types'
+import type { Payload } from './generated-types'
+import { userProfilePayload, API_BASE, UPSERT_ENDPOINT } from '../insider-helpers'
+
+const action: ActionDefinition<Settings, Payload> = {
+  title: 'Update User Profile',
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/insider/updateUserProfile/index.ts:7]:
```suggestion
  title: 'Create or update a user profile',
```

File: packages/destination-actions/src/destinations/insider/insider-helpers.ts [hunk 23]
```diff
@@ -0,0 +1,190 @@
...
+    ]
+  }
+}
+
+export function sendTrackEvent(data: EventPayload) {
+  const payload: upsertUserPayload = {
+    identifiers: {
+      uuid: data.uuid,
+      custom: {
+        segment_anonymous_id: data.segment_anonymous_id
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/insider/insider-helpers.ts:70]: Hi @recep-insider I'm wondering if you should also offer the ability to allow the customer to specify if email and phone should be treated as identifiers for this Action as well?

**Review by joe-ayoub-segment — CHANGES_REQUESTED** [review]
Hi @recep-insider , 
There's a file which accidentally got edited and included in this PR. Could you remove it please?

Also I left some more comments/recommendations. I'll leave it up to you if you want to do anything with them. 

When you respond could you be clear about which comment you've decided to implement and which you decided not to please?

File: packages/destination-actions/src/destinations/insider/insider-helpers.ts (hunk 23 above)
Comment (recep-insider) [reply packages/destination-actions/src/destinations/insider/insider-helpers.ts:70]: Hi @joe-ayoub-segment, yes we will talk with team about that. Probably we will add it.

**Comment by recep-insider** [comment]
@joe-ayoub-segment I added your request but please wait my update for review because we will add one more feature. after that I will update to you for review.

File: packages/destination-actions/src/destinations/insider/trackEvent/index.ts [hunk 24]
```diff
@@ -0,0 +1,380 @@
...
+        product_id: { '@path': '$.properties.product_id' },
+        taxonomy: { '@path': '$.properties.category' },
+        name: { '@path': '$.properties.name' },
+        variant_id: { '@path': '$.properties.variant' },
+        unit_sale_price: { '@path': '$.properties.price' },
+        unit_price: { '@path': '$.properties.price' },
+        quantity: { '@path': '$.properties.quantity' },
+        product_image_url: { '@path': '$.properties.image_url' },
+        event_group_id: {
+          '@if': {
```
Comment (esezerguven) [inline packages/destination-actions/src/destinations/insider/trackEvent/index.ts:128]: Hi @joe-ayoub-segment, I need to add one more property here. order_id cart_id or checkout_id whatever if exists we need to map it as event_group_id. how we can coalesce 3 property ?

**Comment by recep-insider** [comment]
@joe-ayoub-segment hello, we are ready to CR.

**Comment by joe-ayoub-segment** [comment]
Thanks @recep-insider - I'll schedule this for review.

File: packages/destination-actions/src/destinations/insider/trackEvent/index.ts (hunk 24 above)
Comment (esezerguven) [reply packages/destination-actions/src/destinations/insider/trackEvent/index.ts:140]: Hi @joe-ayoub-segment did you have change to check this. I can add this part also before cr

Comment (joe-ayoub-segment) [reply packages/destination-actions/src/destinations/insider/trackEvent/index.ts:146]: Hi @esezerguven , just seeing this now - it was a public holiday yesterday. 
Did you get the answer to your question? 
if not please feel free to schedule a call with me - we won't need the full hour, but we can probably solve your question and get your PR submitted for deployment this Wednesday. 

http://calendly.com/joe_ayoub/integration-workshop

Comment (joe-ayoub-segment) [reply packages/destination-actions/src/destinations/insider/trackEvent/index.ts:146]: I do have a call with Emre tomorrow, if you are able to join that?

**Review by esezerguven — COMMENTED** [review]
image
Hello @joe-ayoub-segment We will add new presets and actions as we talked in the meeting however is this possible to make an preset default disabled and default enable like in the image

**Comment by joe-ayoub-segment** [comment]
Hi @esezerguven , sounds like a good plan. 
Presets will always be enabled by default, and it's not possible to have them pre-configured to be disabled. 
If the customer wants to create a new Mapping then they can click the blue button to create a mapping from an Action or Preset. This is usually sufficient. 

Let me know if you find this too restrictive (and tell me a little more about the use-case) and I might be able to provide some more guidance. 

image

**Comment by joe-ayoub-segment** [comment]
PR review details shared via email. 
Note branch doesn't build and tests don't pass. 
Asked for the 2 separate Destinations to be submitted in separate PRs.

**Comment by esezerguven** [comment]
@joe-ayoub-segment hello joe we have seperate the pull requests. Can you please review here again and inform is there anything. Thanks

**Review by joe-ayoub-segment — APPROVED** [review]
Feedback implemented. 
Tests pass locally. Validate passes.

**Comment by joe-ayoub-segment** [comment]
Hi @emreisik95  this PR has been deployed. I've emailed with follow up steps.

**Pull request merged**